fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/RecursiveUnionCursor.java

```diff
@@ -0,0 +1,264 @@
+                                                   @Nonnull final Supplier<Boolean> isReadingFromInitialCursorSupplier,
+                                                   @Nonnull final Consumer<Boolean> wasInInitialStateConsumer,
+                                                   @Nonnull final Consumer<Boolean> wasReadingFromFirstTempTableConsumer,
+                                                   @Nonnull final Supplier<Boolean> recursiveStepCompletionCallback,
+                                                   @Nonnull Executor executor) {
+        if (unparsed == null) {
+            final var initialCursor = initialCursorCreator.apply(null);
+            final var recursiveCursor = recursiveCursorCreator.apply(null);
+            return new RecursiveUnionCursor<>(initialCursor, () -> recursiveCursorCreator.apply(null), recursiveCursor, executor, isReadingFromInitialCursorSupplier,
+                    recursiveStepCompletionCallback, true);
+        } else {
+            RecordCursorProto.RecursiveCursorContinuation proto;
+            try {
+                proto = RecordCursorProto.RecursiveCursorContinuation.parseFrom(unparsed);
+            } catch (InvalidProtocolBufferException ex) {
+                throw new RecordCoreException("invalid continuation", ex)
+                        .addLogInfo(LogMessageKeys.RAW_BYTES, ByteArrayUtil2.loggable(unparsed));
+            }
+            // resume the state.
+            wasInInitialStateConsumer.accept(proto.getIsInitialState());
+            wasReadingFromFirstTempTableConsumer.accept(proto.getIsReadingInitialCursor());
+            final var initialCursor = initialCursorCreator.apply(proto.getInitialCursorContinuation());
+            final var recursiveCursor = recursiveCursorCreator.apply(proto.getRecursiveCursorContinuation());
+            return new RecursiveUnionCursor<>(initialCursor, () -> recursiveCursorCreator.apply(null), recursiveCursor, executor, isReadingFromInitialCursorSupplier,
+                    recursiveStepCompletionCallback, proto.getIsReadingInitialCursor());
+        }
+    }
+
+    private static class Continuation implements RecordCursorContinuation {
+
+        private final boolean isInitialState;
+
+        @Nonnull
+        private final Supplier<Boolean> isReadingInitialCursorSupplier;
+
+        @Nonnull
+        private final Supplier<RecordCursorContinuation> initialCursorContinuation;
+
+        @Nonnull
+        private final Supplier<RecordCursorContinuation> recursiveCursorContinuation;
+
+        private final boolean isEnd;
+
+        Continuation(boolean isInitialState,
+                     @Nonnull final Supplier<Boolean> isReadingInitialCursorSupplier,
+                     @Nonnull final Supplier<RecordCursorContinuation> initialCursorContinuation,
+                     @Nonnull final Supplier<RecordCursorContinuation> recursiveCursorContinuation,
```

Comment:
Deferring these until `toByteString` when they capture dynamic state makes these delicate. It will not work to capture several results with their continuations and then ask the first to serialize that continuation. This may be acceptable to all callers, but perhaps it would be safer to arrange for it to throw an error.

Reply:
Perhaps temp tables could have a zero-copy snapshot operation that just remembered the current size.

Reply:
Thanks for sharing this valuable insight, I am planning to start working on it in a subsequent PR, considering that the temp table is an append-only data structure, we can reference to multiple snapshots using the a size number only.